src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,139 @@
+                        switchVariable,
+                        context.transformExpression(clause.expression),
+                        lua.SyntaxKind.EqualityOperator
+                    );
+                }
 
-    if (concatenatedIf) {
-        statements.push(concatenatedIf);
-    }
+                // Skip empty clauses
+                if (clause.statements.length === 0) continue;
 
-    const hasDefaultCase = statement.caseBlock.clauses.some(ts.isDefaultClause);
-    statements.push(lua.createGotoStatement(`${switchName}_${hasDefaultCase ? "case_default" : "end"}`));
+                // Declare or assign condition variable
+                statements.push(
+                    isInitialCondition
+                        ? lua.createVariableDeclarationStatement(conditionVariable, condition)
+                        : lua.createAssignmentStatement(
+                              conditionVariable,
+                              lua.createBinaryExpression(conditionVariable, condition, lua.SyntaxKind.OrOperator)
+                          )
+                );
+                isInitialCondition = false;
+            } else {
+                assert(!isInitialCondition, "Default clause should never be the initial condition");
+            }
 
-    for (const [index, clause] of statement.caseBlock.clauses.entries()) {
-        const labelName = `${switchName}_case_${ts.isCaseClause(clause) ? index : "default"}`;
-        statements.push(lua.createLabelStatement(labelName));
-        statements.push(lua.createDoStatement(context.transformStatements(clause.statements)));
-    }
+            // Push if statement for case
+            statements.push(
+                lua.createIfStatement(
+                    conditionVariable,
+                    lua.createBlock(context.transformStatements(clause.statements))
+                )
+            );
+
+            // Clear condition for next clause
+            condition = undefined;
+        }
 
-    statements.push(lua.createLabelStatement(`${switchName}_end`));
+        // Amalgamate the default w/ fallthrough clauses and execute if nothing else executed above
+        const start = clauses.findIndex(c => ts.isDefaultClause(c));
+        if (start >= 0) {
+            const end = statement.caseBlock.clauses
+                .slice(start)
+                .findIndex(clause => containsBreakOrReturn([...clause.statements]));
+            const defaultStatements = statement.caseBlock.clauses
+                .slice(start, end >= 0 ? end + start + 1 : undefined)
+                .reduce<lua.Statement[]>(
+                    (statements, clause) => [...statements, ...context.transformStatements(clause.statements)],
+                    []
+                );
+
+            if (defaultStatements.length) {
+                statements.push(
+                    lua.createIfStatement(
+                        lua.createUnaryExpression(conditionVariable, lua.SyntaxKind.NotOperator),
```

Comment:
Something I was thinking about with this is that in almost all cases, the `if not ____cond` check is unecessary. If code reached this point we definitely want the default clause to run. The one exception is when the last case in the switch is not default and doesn't explicitly break:

```ts
switch (x) {
    default:
        console.log("default");
    case 1:
        console.log(1);
}
```

To compensate for this, we could just add a `break` to the last case if it's not default.

Reply:
So if the default clause was fallen into from the block above, the default clause has already run. I don't want to execute it twice, right?

```ts
switch (1) {
    case 1:
        console.log("1 falls through")
    default:
        console.log("default is triggered and condition is true")
}

// Do not run default again
```

Reply:
Do we need the 2 default case handlers in this example?

Reply:
Oh, I understand now.
We can detect when we need the `not ___cond` default and potentially prune it.

```ts
switch (0) {
    case 1:
        console.log("1 falls through")
    default:
        console.log("default is triggered and condition is true")
    case 3:
        console.log("default falls through")
}
```

You would need the `not ____cond` default in this case.

Reply:
Right, that's the edge case. But what I'm suggesting is instead of doing the `not ____cond`, put a `break` at the end of `case 3`. This would only be necessary if the last case is not default and does not already break itself. At this point I'm getting into less-important optimizations, though 😄